snuba/web/split.py

```diff
@@ -57,40 +90,53 @@ def time_split(dataset, request: Request, *args, **kwargs):
         into smaller increments, and start with the last one, so that we can potentially
         avoid querying the entire range.
         """
+        if not self.__can_execute(request):
+            return None
+
         date_align, split_step = state.get_configs(
             [("date_align_seconds", 1), ("split_step", 3600)]  # default 1 hour
         )
 
-        query_limit = request.query.get_limit()
-        limit = query_limit if query_limit is not None else 0
-        remaining_offset = request.query.get_offset()
-
-        to_date = util.parse_datetime(
-            request.extensions["timeseries"]["to_date"], date_align
+        # We already ensured the conditions below exist in the query during the
+        # can_execute call.
+        to_date_str = next(
+            condition[2]
+            for condition in request.query.get_conditions() or []
+            if _identify_condition(condition, self.__timestamp_col, "<")
         )
-        from_date = util.parse_datetime(
-            request.extensions["timeseries"]["from_date"], date_align
+        to_date = util.parse_datetime(to_date_str, date_align)
+
+        from_date_str = next(
+            condition[2]
+            for condition in request.query.get_conditions() or []
+            if _identify_condition(condition, self.__timestamp_col, ">=")
         )
+        from_date = util.parse_datetime(from_date_str, date_align)
+
+        limit = request.query.get_limit() or 0
```

Comment:
This `or` is redundant with the check in `_is_query_splittable` as well.

Reply:
It is still necessary for mypy not to complain.

Reply:
This also applies to https://github.com/getsentry/snuba/pull/865/files#r410393313 but writing it here since this is a bit more self-contained an example. This seems more of a reflection of the structure of this method and the duplicated logic in relationship to `__can_execute` and `_is_query_splittable` than of deficiencies in the type checker.

If `execute` was structured this way, this wouldn't be an issue:

```python
def execute(…) -> Optional[RawQueryResult]:
  # establish invariants (basically inline __can_execute and _is_query_splittable)
  limit = request.query.get_limit()
  if limit is None:
    return None

  to_date = …
  if to_date is None:
    return None

  # run split
  … everything else already in here …
```

This was the point of this comment but I guess it was more implied than directly called out: https://github.com/getsentry/snuba/pull/865#discussion_r404557527